core/src/test/java/org/testcontainers/containers/wait/strategy/WaitAllStrategyTest.java

```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    static class SleepingStrategy extends AbstractWaitStrategy {
+
+        private final long sleepingDuration;
+
+        SleepingStrategy(Duration defaultInnerWait) {
+            super.startupTimeout = defaultInnerWait;
+            // Always oversleep by default
+            this.sleepingDuration = defaultInnerWait.multipliedBy(2).toMillis();
+        }
+
+        @Override
+        protected void waitUntilReady() {
+            try {
+                // Don't use ducttape, make sure we don't catch the wrong TimeoutException later on.
+                CompletableFuture.runAsync(
+                    () -> Uninterruptibles.sleepUninterruptibly(this.sleepingDuration, TimeUnit.MILLISECONDS)
+                ).get((int) startupTimeout.toMillis(), TimeUnit.MILLISECONDS);
+            } catch (InterruptedException | ExecutionException e) {
+            } catch (java.util.concurrent.TimeoutException e) {
+                throw new CaughtTimeoutException("Inner wait timed out, outer strategy didn't apply.");
+            }
+        }
+    }
+
+    @Test
+    public void appliesOuterTimeoutTooLittleTime() {
+
+        Duration defaultInnerWait = Duration.ofMillis(2);
+        Duration outerWait = Duration.ofMillis(6);
+
+        WaitStrategy child1 = new SleepingStrategy(defaultInnerWait);
+        WaitStrategy child2 = new SleepingStrategy(defaultInnerWait);
+
+        final WaitStrategy underTest = new WaitAllStrategy()
+            .withStrategy(child1)
+            .withStrategy(child2)
+            .withStartupTimeout(outerWait);
+
+        assertThrows("The outer strategy timeout applies", TimeoutException.class, () -> {
+            underTest.waitUntilReady(container);
+        });
+    }
+
+    @Test
+    public void appliesOuterTimeoutEnoughTime() {
+
+        Duration defaultInnerWait = Duration.ofMillis(2);
+        Duration outerWait = Duration.ofMillis(20);
+
+        WaitStrategy child1 = new SleepingStrategy(defaultInnerWait);
+        WaitStrategy child2 = new SleepingStrategy(defaultInnerWait);
+
+        final WaitStrategy underTest = new WaitAllStrategy()
+            .withStrategy(child1)
+            .withStrategy(child2)
+            .withStartupTimeout(outerWait);
+
+        try {
+            underTest.waitUntilReady(container);
+        } catch (Exception e) {
+            if (e.getCause() instanceof CaughtTimeoutException)
+                fail("The timeout wasn't applied to inner strategies.");
+            else {
+                fail(e.getMessage());
+            }
+        }
+    }
+
+    @Test
+    public void appliesOuterTimeoutOnAdditionalChildren() {
+
+        Duration defaultInnerWait = Duration.ofMillis(2);
+        Duration outerWait = Duration.ofMillis(20);
+
+        WaitStrategy child1 = new SleepingStrategy(defaultInnerWait);
+        WaitStrategy child2 = new SleepingStrategy(defaultInnerWait);
+
+        final WaitStrategy underTest = ((WaitAllStrategy) new WaitAllStrategy()
```

Comment:
this test demonstrates the usability issue in `WaitAllStrategy`, we should change `withStartupTimeout`'s return method to `WaitAllStrategy`. You can do it in this PR if you want :)

/cc @rnorth